### PR TITLE
Proposals system. Version 2. Iteration6

### LINKED ISSUE #174: Map Substrate origin parameter to the MemberId and CouncillorId in the proposals module
In the **proposal modules** we need to implement mapping from the Substrate origin parameter in extrinsics to the _actor_id_.

For example in the `create_proposal` API call, `actor_id` is `proposer_id`:

```
        let account_id = T::ProposalOrigin::ensure_origin(origin)?;
        let proposer_id = T::ProposerId::from(account_id.clone());
```

We need to implement these conversions in the **engine module** for ProposerId and VoterId and in the **discussion module** for the ThreadAuthorId and PostAuthorId.

### LINKED ISSUE #226: Proposasls Iteration 6 fixes after PR
Parent [Iteration 6 PR](https://github.com/Joystream/substrate-runtime-joystream/pull/194)

Proposals iteration 6 fixes:
- rename `proposal_code` to the DispatchableCode (and related parameters to 'encoded_')
- move actor validators to the runtime
- rename `ProposalObject`
- rename `StakeData` to the ActiveStake
- move ActiveState(`StakeData`) - to the 'Active' proposal status at first, and then to the 'Finalized'
- refactor `execute_proposal` method
- rename `finalization_error` to the encoded_unstaking_error_due_to_broken_runtime
- move actor validation checks to the codex from the `engine` and `discussion` modules
- move out ensure_ methods of the engine and discussion to the codex

### PR DESCRIPTION
**Pull request by shamil-gadelshin**
### Features
- monorepo integration
- introduced `ActorOriginValidator` in the common module for future integration with the membership and council modules
- proposal system authentication system (membership integration) in the engine and discussion modules (`ActorOriginValidator` implementation - `MembershipOriginValidator`)
- council integration for voting: `ActorOriginValidator` and `VotersParameters` implementations with tests - `CouncilManager`
- created tests for `MembershipValidator` and `CouncilManager`
- returned stake after the unstaking (StakingEventsHandler integration)
- removed proposals v1
- runtime integration and implementing all handlers required (council handler, total voters handler)
- change proposals execution model
- move all proposals modules from string error constants to decl_error! approach
- jammed 'srml_staking_reward_curve::build!' [warning](https://github.com/Joystream/substrate-runtime-joystream/issues/179)
- `StakingEventsHandler` integration with the `ContentWorkingGroupStakingEventHandler`
- `set_countil()` is public now for testing purposes (it's guarded by `ensure_root()`))

### Related issues

- https://github.com/Joystream/substrate-runtime-joystream/issues/174
- https://github.com/Joystream/substrate-runtime-joystream/issues/179
- https://github.com/Joystream/substrate-runtime-joystream/issues/154

### [Iteration 5 features (was skipped)](https://github.com/Joystream/substrate-runtime-joystream/pull/176)
- spam gating: prevent the same author from writing more than N posts in a row.
- change `EnsureOrigin` to the `ActorOriginValidator` model (was rewritten during the 6th iteration )
- create events for the discussion system

### CODE REVIEW

**Review by bedeho — CHANGES_REQUESTED** [review]
- This review goes beyond the diff in the PR, because it was frankly really hard to just follow that, given how long ago I looked at this, and also because prior reviews did not go that in depth on different extrinsics, as we were just getting started.

- In general, types with postfix "data" indicate a sloppy name or bad type, unless they literally represent unstructured data without any internal meaning. There is at least `StakeData` and `FinalizationData` in this iteration. For the former, I think it should not exist (as you will see in the review), and for the latter, see next section.

- Question: Is there any reason to check origin on pure `Module` methods? doesnet seem to make sense? the calling code will perform the origin check. Perhaps this is why we have the akward double origin thing... its not needed?

# Engine

- When a council changed, I think we agreed to reset all the current votes and reset the finalization deadline. That is currently not implemented, right?

- `execute_proposal` appears to have a bug, as it unconditionally removes from `PendingExecutionProposalIds`, even if the first guard rejects the proposal. Moreover, the routine does not check that the proposal actually exists, yet it does check that it has the right state. This seems a bit inconsistent. I think that if its going to make an assumption without checking, it needs not check for any even weaker assumptions. Moreover, its also a bit akward that it is rechecking whether the given proposal is finalized, despite the caller already having done this check in computation `get_approved_proposal_with_expired_grace_period_ids`, that should be avoidable in the same code path by picking the right representations. Have a look, it may not be worth changing so much, but I wanted to draw your attention to it.

- For `FinalizationData`
  - I think there is a crisper name could be `FinalizedProposalStatus`
  - The field `finalization_error` should reflect that this is specifically the serialized error associated with unstaking. When reading it I first thought this is what held the error related to execution as well, but that is indeed separate, so name should be sharpened. Also, something needs to be done to very clearly signal that this is no normal error, but indeed the broken invariant error. At the very least the comment must say so, but perhaps also the name, e.g. `encoded_unstaking_error_due_to_broken_runtime`. Its a bit long, but the length helps to grab the attention to how distinct this field is.
  - Lets drop `finalized_at`. Its purely archival, and so query node should hold it, and it's not even clear that we will, in the end, hold the entire proposal itself after it is executed.

- Did we not decide that, for the time being, we will not store broken invariant errors in storage, simply for the benefit of archiving error messages. We would use either console logging or something else to capture them? Unfortunately, we should have written down what we decided. At any rate, this sort of storing of errors seems to do little if anything of increasing safety or blocking further state corruption, which are the strong reasons for handling broken runtimes in some way. 

- The only reason we have `ActiveProposalIds` is so that more effectively iterate only the proposals that are active when trying to finalize. This also why we need to limit how many can be active at once.

  This is as intended, but I want to explore an alternative. If you think its an improvement, lets make an issue out of it and have someone work on it at a later time:

  We drop `ActiveProposalIds` and the limit on the number of active proposals, as well as any iteration in `on_finalized`. Instead, a new map `finalize_at_block: T::BlockNumber => T::ProposalId` is introduced, and a mapping is introduced whenever a proposal is created. This map is checked every `on_finalized`, but there is no iteration, you just check for the specific block you are on. There is an implicit limit that only one proposal can be finalized per block, and that is enforced upon proposal creation. 

  It seems this is both safer, as there is no enumeration and iteration, and we drop one parameter value, and we have no limit on the number of simultaneous proposals in flight.

  What do you think?

  The same would apply to `PendingExecutionProposalIds`.

- Seems like this storage map should be dropped, and instead just hold the value as a field in the `struct Proposal`.

  `pub ProposalCode get(fn proposal_codes): map T::ProposalId =>  Vec<u8>;`

# Codex

Looks great!

# Discussion system

Looks great!

File: runtime-modules/proposals/engine/src/lib.rs [hunk 1]
```diff
@@ -482,69 +572,93 @@ impl<T: Trait> Module<T> {
...
     <T as system::Trait>::BlockNumber,
-    <T as Trait>::ProposerId,
+    MemberId<T>,
     types::BalanceOf<T>,
     <T as stake::Trait>::StakeId,
+    <T as system::Trait>::AccountId,
 >;
 
 // Simplification of the 'Proposal' type
 type ProposalObject<T> = Proposal<
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply runtime-modules/proposals/engine/src/lib.rs]: Agree, will change to 'ProposalOf<T>' similar with BalanceOf<T>

File: runtime-modules/proposals/engine/src/lib.rs [hunk 2]
```diff
@@ -91,22 +97,24 @@ pub trait Trait: system::Trait + timestamp::Trait + stake::Trait {
 
     /// Defines max simultaneous active proposals number.
     type MaxActiveProposalLimit: Get<u32>;
+
+    /// Proposals executable code. Can be instantiated by external module Call enum members.
+    type ProposalCode: Parameter + Dispatchable<Origin = Self::Origin> + Default;
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply runtime-modules/proposals/engine/src/lib.rs]: Ok.

**Comment by shamil-gadelshin** [comment]
> When a council changed, I think we agreed to reset all the current votes and reset the finalization deadline. That is currently not implemented, right?

We didn't discuss it yet. I will add it to my backlog.

**Comment by shamil-gadelshin** [comment]
We need `finalized_at` not only for archival purposes, but we calculate our `grace_period` using its value.

**Comment by shamil-gadelshin** [comment]
> This map is checked every on_finalized, but there is no iteration, you just check for the specific block you are on. There is an implicit limit that only one proposal can be finalized per block, and that is enforced upon proposal creation.

Proposal veto and cancellation cannot be predicted

File: runtime-modules/proposals/engine/src/types/mod.rs [hunk 3]
```diff
@@ -139,11 +152,12 @@ pub struct Proposal<BlockNumber, ProposerId, Balance, StakeId> {
     /// Curring voting result for the proposal
     pub voting_results: VotingResults,
 
-    /// Created stake id for the proposal
-    pub stake_id: Option<StakeId>,
+    /// Stake data for the proposal
+    pub stake_data: Option<StakeData<StakeId, AccountId>>,
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply runtime-modules/proposals/engine/src/types/mod.rs]: Yes, we were going to decide whether to remove stake information from the system on proposal finalization.

File: runtime-modules/proposals/engine/src/lib.rs [hunk 4]
```diff
@@ -20,60 +20,66 @@
...
+use membership::origin_validator::MemberId;
+use srml_support::dispatch::Dispatchable;
 
 /// Proposals engine trait.
-pub trait Trait: system::Trait + timestamp::Trait + stake::Trait {
+pub trait Trait:
+    system::Trait
+    + timestamp::Trait
+    + stake::Trait
+    + membership::members::Trait
```
[the remark this answers is not in the document]
Comment (shamil-gadelshin) [reply runtime-modules/proposals/engine/src/lib.rs]: `Governance` module dependency is needed to create `CouncilManager`. We can move it to the `council` module, but then the `governance` module will be dependent upon the `membership` module.

**Comment by shamil-gadelshin** [comment]
Work to be done using this [issue](https://github.com/Joystream/substrate-runtime-joystream/issues/226)

**Review by bedeho — APPROVED** [review]
Great job!

**Pull request merged**